etcdserver/server.go

```diff
@@ -689,6 +689,12 @@ func (s *EtcdServer) run() {
 	}
 
 	defer func() {
+		// stop accepting new connections, RPCs,
+		// and blocks until all pending RPCs are finished
+		if s.Cfg != nil && s.Cfg.OnShutdown != nil {
+			s.Cfg.OnShutdown()
```

Comment:
I think it's possible to avoid having this in Cfg entirely-- this function could be called prior to calling HardStop/Stop; similar to how the listeners are closed in embed.Etcd.Close() before calling `Server.Stop()`